### PR TITLE
Restore `restore-file` feature

### LINKED ISSUE #6155: `restore-file` no longer works
### Description

I guess since GitHub shipped a lot of features while GitHub Universe the "restore file" feature from a pull requests file-context menu no longer works.

I get a

> can't access property "elements", form is undefined

error when using 

grafik

### How to replicate the issue + URL

https://github.com/FriendsOfREDAXO/rexstan/pull/225

### Extension version

22.11.6

### Browser(s) used

firefox 108 on latest macOS ventura

### PR DESCRIPTION
**Pull request by gamemaker1**
### Related Issues

Closes #6155.

## Test URLs

https://github.com/gamemaker1/test/pull/1/files

### CODE REVIEW

**Comment by fregante** [comment]
Did you verify whether this works on PR from forks? This was the issue last time we tried

**Comment by fregante** [comment]
This doesn't work when the file contains non-ascii characters unfortunately:

> Failed to execute 'btoa' on 'Window': The string to be encoded contains characters outside of the Latin1 range.

https://user-images.githubusercontent.com/1402241/203085877-30e5e7d9-24a6-4c6c-af36-8a0cef797329.mov

See https://stackoverflow.com/a/54302528

I think you should be able to test it on this PR, once fixed, exactly like in my demo. If the file is restored correctly, then we can merge it!

**Comment by fregante** [comment]
I'm confused as to why we're getting a raw text string and pushing a base64. There must be a way to avoid having to do any conversion. In my first iteration, I was using the API v3 for this, but I had to turn to `pushForm` for the same issue we had on your last PR:

- [`7e8bb26` (#2182)](https://github.com/refined-github/refined-github/pull/2182/commits/7e8bb264cda190a5ce2e09e51196cb8f4b230696)

**Comment by gamemaker1** [comment]
> Did you verify whether this works on PR from forks? This was the issue last time we tried

Yes, it works - https://github.com/dabbu-knowledge-platform/test/pull/1

**Comment by gamemaker1** [comment]
I don't see the Restore File button for images :/

image

**Comment by gamemaker1** [comment]
Turns out its because the "Edit file" button is disabled ([L120](https://github.com/refined-github/refined-github/pull/6171/files#diff-8041b02476782c5d40a59fb4fad7627e7d6acb7e257bb46e5bd7e8ad4ee1b40bR120)), which makes sense to me.

**Comment by gamemaker1** [comment]
I can't play the video you shared:

image

But it does seem to work with non-ascii characters: https://github.com/dabbu-knowledge-platform/test/pull/1/commits/776b841720d9449448767d70e5c0fba163b0d5f9

**Comment by fregante** [comment]
The non-ASCII files need to be in the restored version, not the "new" one. Try on this PR.

As for the video, got can download it and open it with VLC if the browser won't play it

**Comment by fregante** [comment]
@gamemaker1 can you finalize this PR? I'd love to ship it this week

**Comment by gamemaker1** [comment]
Done :)

**Comment by fregante** [comment]
Neat! I think it works, tested in Chrome and Firefox: https://github.com/refined-github/sandbox/pull/54


Will merge and release a new version soon

**Review by fregante — APPROVED** [review]

File: source/features/restore-file.tsx [hunk 1]
```diff
@@ -115,10 +140,10 @@ function handleMenuOpening({delegateTarget: dropdown}: DelegateEvent): void {
...
 }
 
-function init(signal: AbortSignal): void {
-	// `capture: true` required to be fired before GitHub's handlers
-	delegate(document, '.file-header .js-file-header-dropdown', 'toggle', handleMenuOpening, {capture: true, signal});
-	delegate(document, '.rgh-restore-file', 'click', handleRestoreFileClick, {capture: true, signal});
+function init(): void {
+	// `useCapture` required to be fired before GitHub's handlers
+	delegate(document, '.file-header .js-file-header-dropdown', 'toggle', handleMenuOpening, true);
+	delegate(document, '.rgh-restore-file', 'click', handleRestoreFileClick, true);
```
Comment (fregante) [inline source/features/restore-file.tsx:146]: This change should be reverted

File: source/features/restore-file.tsx [hunk 2]
```diff
@@ -53,42 +60,60 @@ async function restoreFile(progress: (message: string) => void, menuItem: Elemen
...
-	menuItem.closest('.file')!.remove();
+		// Show toast while restoring
+		await showToast(async progress => restoreFile(progress!, menuItem, filePath), {
+			message: 'Restoring…',
+			doneMessage: 'Restored!',
+		});
+
+		// Hide file from view
+		menuItem.closest('.file')!.remove();
+	} catch (error: unknown) {
```
Comment (fregante) [inline source/features/restore-file.tsx:114]:
```suggestion
	} catch (error) {
```

File: source/features/restore-file.tsx [hunk 3]
```diff
@@ -53,42 +60,60 @@ async function restoreFile(progress: (message: string) => void, menuItem: Elemen
...
 		const filePath = menuItem.closest<HTMLDivElement>('[data-path]')!.dataset.path!;
-		return restoreFile(progress!, menuItem, filePath);
-	}, {
-		message: 'Restoring…',
-		doneMessage: 'Restored!',
-	});
-
-	// Hide file from view
-	menuItem.closest('.file')!.remove();
+		// Show toast while restoring
```
Comment (fregante) [inline source/features/restore-file.tsx:106]:
```suggestion
```

File: source/features/restore-file.tsx [hunk 4]
```diff
@@ -53,42 +60,60 @@ async function restoreFile(progress: (message: string) => void, menuItem: Elemen
...
 async function handleRestoreFileClick(event: DelegateEvent<MouseEvent, HTMLButtonElement>): Promise<void> {
 	const menuItem = event.delegateTarget;
 
-	await showToast(async progress => {
+	// Only allow one click
+	if (filesRestored.has(menuItem)) {
+		return;
+	}
+
+	filesRestored.add(menuItem);
```
Comment (fregante) [inline source/features/restore-file.tsx:102]: We can drop this too. It's part of the old logic that kept the menu open to display the status. Thanks to `showToast`, double-clicks are impossible now

**Comment by gamemaker1** [comment]
Oh I'm sorry I couldn't make the changes quickly, I just saw the review comments :(

**Pull request merged**